satellite/satellitedb/certdb.go

```diff
@@ -18,40 +19,47 @@ type certDB struct {
 	db *dbx.DB
 }
 
-func (b *certDB) SavePublicKey(ctx context.Context, nodeID storj.NodeID, publicKey crypto.PublicKey) error {
-	tx, err := b.db.Open(ctx)
+func (certs *certDB) SavePublicKey(ctx context.Context, nodeID storj.NodeID, publicKey crypto.PublicKey) error {
+	_, err := certs.db.Get_CertRecord_By_Id(ctx, dbx.CertRecord_Id(nodeID.Bytes()))
+	if err == sql.ErrNoRows {
+		return certs.tryAddPublicKey(ctx, nodeID, publicKey)
+	}
 	if err != nil {
 		return Error.Wrap(err)
 	}
 
-	_, err = tx.Get_CertRecord_By_Id(ctx, dbx.CertRecord_Id(nodeID.Bytes()))
+	// nodeID entry already exists, just return
+	return nil
+}
+
+func (certs *certDB) tryAddPublicKey(ctx context.Context, nodeID storj.NodeID, publicKey crypto.PublicKey) error {
+	// no rows err, so create/insert an entry
+	pubbytes, err := pkcrypto.PublicKeyToPKIX(publicKey)
 	if err != nil {
-		// no rows err, so create/insert an entry
-		pubbytes, err := pkcrypto.PublicKeyToPKIX(publicKey)
-		if err != nil {
-			return Error.Wrap(errs.Combine(err, tx.Rollback()))
-		}
-
-		_, err = tx.Create_CertRecord(ctx,
-			dbx.CertRecord_Publickey(pubbytes),
-			dbx.CertRecord_Id(nodeID.Bytes()),
-		)
-		if err != nil {
-			return Error.Wrap(errs.Combine(err, tx.Rollback()))
-		}
-	} else {
-		// nodeID entry already exists, just return
-		return Error.Wrap(tx.Rollback())
+		return Error.Wrap(err)
 	}
 
-	return Error.Wrap(tx.Commit())
+	// TODO: use upsert here instead of create
+	_, err = certs.db.Create_CertRecord(ctx,
+		dbx.CertRecord_Publickey(pubbytes),
+		dbx.CertRecord_Id(nodeID.Bytes()),
+	)
+	// another goroutine might race to create the cert record, let's ignore that error
+	if pgutil.IsConstraintError(err) || sqliteutil.IsConstraintError(err) {
```

Comment:
might make sense to have a general dbutil.IsConstraintError so we don't have to find all these places if we add support for spanner or mysql or something

Reply:
The reason I didn't want to add `dbutil.IsConstraintError` is because that would most likely end up pulling in additional dependencies. (e.g. storage node depending on postgres)

Reply:
Oh yeah. Thats a good point Egon.